core/src/main/java/com/linecorp/armeria/common/logging/RequestScopedMdc.java

```diff
@@ -224,24 +284,28 @@ public static void copy(RequestContext ctx, String key) {
     public static void copyAll(RequestContext ctx) {
         requireNonNull(ctx, "ctx");
         checkState(delegate != null, ERROR_MESSAGE);
-
-        final Map<String, String> map = delegate.getCopyOfContextMap();
-        if (map == null || map.isEmpty()) {
-            return;
+        final Map<String, String> map = getDelegateContextMap();
+        if (map != null) {
+            putAll(ctx, map);
         }
+    }
 
-        synchronized (ctx) {
-            final Map<String, String> oldMap = getMap(ctx);
-            final Map<String, String> newMap;
-            if (oldMap.isEmpty()) {
-                newMap = map;
-            } else {
-                newMap = new Object2ObjectOpenHashMap<>(oldMap.size() + map.size());
-                newMap.putAll(oldMap);
-                newMap.putAll(map);
+    @Nullable
+    private static Map<String, String> getDelegateContextMap() {
+        assert delegate != null;
+        try {
+            // Try to use `LogbackMDCAdapter.getPropertyMap()` which does not make a copy.
+            @SuppressWarnings("unchecked")
+            final Map<String, String> map =
+                    delegateGetPropertyMap != null ? (Map<String, String>) delegateGetPropertyMap.invokeExact()
+                                                   : delegate.getCopyOfContextMap();
+            if (map != null) {
+                return map;
             }
-            ctx.setAttr(MAP, Collections.unmodifiableMap(newMap));
+        } catch (Throwable t) {
+            Exceptions.throwUnsafely(t);
```

Comment:
Isn't it better to log (maybe once?) and return `delegate.getCopyOfContextMap();` in this case?

Reply:
We will never reach here, because we made sure `invokeExact()` works in the class initializer. If any exception is raised, it means it's a bug. Let me add some comment about this.